### PR TITLE
RC: VBR (Mode 2) and CVBR (Mode 3) 

### PR DESCRIPTION
**Pull request by anaghdin**
VBR (Mode 2) and CVBR (Mode 3) integrated from SVT-VP9
@hassount : Please check the increase in number of PCS based on the lookahead size. added to fix the deadlock

### CODE REVIEW

**Comment by j-mas** [comment]
Encoding with rc 2 always stopped at frame 1982  !!!

Number of logical cores available: 28
Number of PPCS 45
-------------------------------------------
SVT [config]: Main Profile      Tier (auto)     Level (auto)
SVT [config]: EncoderMode                                                       : 7
SVT [config]: EncoderBitDepth / CompressedTenBitFormat                          : 8 / 0
SVT [config]: SourceWidth / SourceHeight                                        : 1920 / 1080
SVT [config]: Fps_Numerator / Fps_Denominator / Gop Size / IntraRefreshType     : 24000 / 1001 / 16 / 1
SVT [config]: HierarchicalLevels / BaseLayerSwitchMode / PredStructure          : 3 / 0 / 2
SVT [config]: RCMode / TargetBitrate / LookaheadDistance / SceneChange          : VBR / 1100000 / 15 / 0
-------------------------------------------
Encoding      1982
C:\>

**Comment by hassount** [comment]
@anaghdin 
Can you please fix that linux warnings ?

**Comment by j-mas** [comment]
@anaghdin
What happened? With the latest AppVeyor build 1.0.295 the VBR (Mode 2) and CVBR (Mode 3) is deleted. I have to use build 1.0.292

File: Config/Sample.cfg [hunk 1]
```diff
@@ -82,5 +81,5 @@ AsmType                         : 1             # Assembly instruction set (0: L
 LogicalProcessors               : 0             # The number of logical processor which encoder threads run on [0-N] (N is maximum number of logical processor)
 TargetSocket                    : -1            # For dual socket systems, this can specify which socket the encoder runs on (-1=Both Sockets, 0=Socket 0, 1=Socket 1)
 #====================== Rate Control ===============================
-RateControlMode                 : 0             # Rate control mode (0: OFF(CQP), 1: ABR)
+RateControlMode                 : 0             # Rate control mode (0: OFF(CQP), 1: ABR, 2: VBR)
```
Comment (hassount) [inline Config/Sample.cfg]: can we add 3: CVBR

File: Source/Lib/Codec/EbEntropyCodingProcess.c [hunk 2]
```diff
@@ -593,6 +597,19 @@ void* EntropyCodingKernel(void *input_ptr)
                         sb_ptr,
                         picture_control_set_ptr);
 
+#if RC            
+                    sb_ptr->total_bits = 0;
+                    uint32_t prev_pos = sb_index ? picture_control_set_ptr->entropy_coder_ptr->ecWriter.ec.offs: 0;//residual_bc.pos
+                    EbPictureBufferDesc_t *coeffPicturePtr = sb_ptr->quantized_coeff;
```
Comment (hassount) [inline Source/Lib/Codec/EbEntropyCodingProcess.c]: snake_case

File: Source/Lib/Codec/EbMotionEstimationProcess.c [hunk 3]
```diff
@@ -691,15 +693,19 @@ void* MotionEstimationKernel(void *input_ptr)
 
                         if (sb_width == BLOCK_SIZE_64 && sb_height == BLOCK_SIZE_64) {
 
+#if RC 
 
+                            intra_sad_interval_index = picture_control_set_ptr->variance[sb_index][ME_TIER_ZERO_PU_64x64] >> 4;
+#else
                             //DOUBLE CHECK THIS PIECE OF CODE
+                            uint32_t                       bestOisCuIndex = 0;
```
Comment (hassount) [inline Source/Lib/Codec/EbMotionEstimationProcess.c]: snake_case

File: Source/Lib/Codec/EbMotionEstimationProcess.c [hunk 4]
```diff
@@ -691,15 +693,19 @@ void* MotionEstimationKernel(void *input_ptr)
 
                         if (sb_width == BLOCK_SIZE_64 && sb_height == BLOCK_SIZE_64) {
 
+#if RC 
 
+                            intra_sad_interval_index = picture_control_set_ptr->variance[sb_index][ME_TIER_ZERO_PU_64x64] >> 4;
+#else
                             //DOUBLE CHECK THIS PIECE OF CODE
```
Comment (hassount) [inline Source/Lib/Codec/EbMotionEstimationProcess.c]: Do we still need this comment ?

File: Source/Lib/Codec/EbPacketizationProcess.c [hunk 5]
```diff
@@ -81,7 +84,190 @@ static void write_td (
...
+
+            uint64_t ref_qindex_dequant = (uint64_t)picture_control_set_ptr->parent_pcs_ptr->deq.y_dequant_QTX[picture_control_set_ptr->parent_pcs_ptr->base_qindex][1];
+            uint64_t sad_bits_ref_dequant = 0;
+            uint64_t weight = 0;
+            {
+                if (picture_control_set_ptr->slice_type == I_SLICE) {
+                    if (sequence_control_set_ptr->input_resolution < INPUT_SIZE_4K_RANGE) {
+                        for (sad_interval_index = 0; sad_interval_index < NUMBER_OF_INTRA_SAD_INTERVALS; sad_interval_index++) {
+
+                            if (count[sad_interval_index] > 5) {
```
Comment (hassount) [inline Source/Lib/Codec/EbPacketizationProcess.c]: single-line statements do not need brackets, ditto below

**Comment by hassount** [comment]
Rebased with #193

**Pull request closed without merging**